storage-node/packages/runtime-api/index.js

```diff
@@ -84,225 +91,247 @@ class RuntimeApi {
-        debug(`Got nonce for ${accountId} from chain: ${nonce}`)
+      lastTxUpdateResult = result
+
+      // Deal with statuses which will prevent
+      // extrinsic from finalizing.
+      if (status.isUsurped) {
+        debug(status.type)
+        debug(JSON.stringify(status.asUsurped))
+        onFinalizedFailed && onFinalizedFailed({ err: 'Usurped', result, tx: status.asFinalized })
       }
 
-      return new Promise((resolve, reject) => {
-        debug('Signing and sending tx')
-        // send(statusUpdates) returns a function for unsubscribing from status updates
-        const unsubscribe = tx
-          .sign(fromKey, { nonce })
-          .send(({ events = [], status }) => {
-            debug(`TX status: ${status.type}`)
-
-            // Whatever events we get, process them if there's someone interested.
-            // It is critical that this event handling doesn't prevent
-            try {
-              if (subscribed && callback) {
-                const matched = RuntimeApi.matchingEvents(subscribed, events)
-                debug('Matching events:', matched)
-                if (matched.length) {
-                  callback(matched)
-                }
-              }
-            } catch (err) {
-              debug(`Error handling events ${err.stack}`)
-            }
+      if (status.isDropped) {
+        debug(status.type)
+        debug(JSON.stringify(status.asDropped))
+        onFinalizedFailed && onFinalizedFailed({ err: 'Dropped', result, tx: status.asFinalized })
+      }
 
-            // We want to release lock as early as possible, sometimes Ready status
-            // doesn't occur, so we do it on Broadcast instead
-            if (status.isReady) {
-              debug('TX Ready.')
-              incrementNonce()
-              resolve(unsubscribe) // releases lock
-            } else if (status.isBroadcast) {
-              debug('TX Broadcast.')
-              incrementNonce()
-              resolve(unsubscribe) // releases lock
-            } else if (status.isFinalized) {
-              debug('TX Finalized.')
-              finalizedPromise.resolve(status)
-            } else if (status.isFuture) {
-              // comes before ready.
-              // does that mean it will remain in mempool or in api internal queue?
-              // nonce was set in the future. Treating it as an error for now.
-              debug('TX Future!')
-              // nonce is likely out of sync, delete it so we reload it from chain on next attempt
-              delete this.nonces[accountId]
-              const err = new Error('transaction nonce set in future')
-              finalizedPromise.reject(err)
-              reject(err)
-            }
+      // When is this status emitted?
+      if (status.isInvalid) {
+        debug(status.type)
+        debug(JSON.stringify(status.asInvalid))
+        onFinalizedFailed && onFinalizedFailed({ err: 'Invalid', result, tx: status.asFinalized })
```

Comment:
Can we use `asFinalized` with `isInvalid` without causing an error?

Reply:
fixed in https://github.com/Joystream/joystream/pull/1063/commits/cd1bbb859081064b603dc7fdec7563e2d8dced3d and then in https://github.com/Joystream/joystream/pull/1063/commits/a779bc51f9b2ff2d18d7381471891367e5e60870